# Desafio: Construindo um modelo de Regressão para Marketing

# Contexto - Introdução

Uma empresa está investindo mensalmente em plataformas de publicidade online, como Youtube, Facebook e newspaper, para a prospecção de leads (pessoas interessadas em seus produtos). A fim de acompanhar o desempenho desses investimentos, a empresa registra todos os gastos com publicidade e todos os retornos de vendas gerados a partir desses investimentos.

Para **entender** melhor **a relação entre as variáveis** presentes nesses registros e **identificar os fatores que mais impactam** na geração de leads, a empresa solicitou a análise de um especialista em dados. **Além disso, a empresa busca criar um modelo de predição** de valores para estimar o retorno de vendas que pode ser gerado a partir de um determinado investimento em publicidade.

# Sobre os dados

A tabela contém informações dos investimentos feitos pelo youtube, facebook, newspaper e também a quantidade de cada. 

| Coluna    | Descrição              |
|-----------|------------------------|
| youtube   | Investimento youtube   |
| facebook  | Investimento facebook  |
| newspaper | Investimento newspaper |
| sales     | Valor das vendas       |

# Importando as Bibliotecas

In [19]:
#Manipulação de Dados
import pandas as pd
import numpy as np
import math

#Visualização
import plotly.express as px
import plotly.subplots as sp
import plotly.graph_objs as go
import plotly.figure_factory as ff
from plotly.subplots            import make_subplots

#Modelo
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

#Display
import warnings
import handcalcs.render
from IPython.display            import Markdown

# Funções Auxiliares e Configurações

In [2]:
warnings.filterwarnings("ignore")
pd.options.display.float_format = '{:.2f}'.format

def df_informations(df):
    # Cria um Dataframe com as informações relativas ao Dataset
    df_info = pd.DataFrame({'Not Null': df.notnull().sum(),
                            'Null': df.isnull().sum(),
                            'Perce Null': df.isnull().sum() / len(df),
                            'Dtype': df.dtypes
                })
    
    # Cria outro Dataframe com todos os types do df
    df_dtype = df_info['Dtype'].value_counts().reset_index()
    df_dtype.columns = ['Dtype', 'Count']
    df_dtype['Perce'] = round(df_dtype['Count'] / df_dtype['Count'].sum(), 2)
    
    # Verificando se tem linhas duplicadas
    duplicado = df.duplicated().sum()
    mensagem = 'Não possui linhas duplicadas.' if duplicado == 0 else f"Tem {duplicado} linhas duplicadas, o que representa {round(duplicado/len(df)*100, 2)}% do total de linhas."

    # Texto automático das dimensões do df
    text = f'Dataset tem {df.shape[0]} linhas e {df.shape[1]} colunas. {mensagem} Sobre o dataset, temos:'

    # Colore os Dataframes
    df_info = df_info.style.background_gradient(cmap='jet', subset=['Perce Null']).format({'Perce Null': '{:.2%}'})
    df_dtype = df_dtype.style.background_gradient(cmap='YlGn', subset=['Perce']).format({'Perce': '{:.2%}'})

    # Texto final com Markdown
    display(Markdown("<H3 style='text-align:left;float:lfet;'>Informações sobre o Dataset"))
    display(Markdown(f'<H5> {text}'))
    display(df_info)
    display(Markdown("<H3 style='text-align:left;float:lfet;'>Sobre Dtypes, temos:"))
    display(df_dtype)
    display(Markdown("<H3 style='text-align:left;float:lfet;'>Estatística Descritiva"))
    display(df.describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99]).T)

def plot_distribution_and_boxplot(df):
    # Obtém a lista de colunas numéricas
    num_cols = df.select_dtypes(include=[np.number]).columns
    
    # Cria a lista de títulos para os subplots
    subplot_titles = [title for col in num_cols for title in [f"Histograma - {col}", f"Boxplot - {col}"]]
    
    # Cria a figura com subplots
    fig = sp.make_subplots(rows=len(num_cols), cols=2, subplot_titles=subplot_titles)
    
    # Define cores
    colors = ['blue', 'red', 'green', 'purple', 'orange', 'brown', 'pink', 'yellow', 'cyan', 'magenta', 
              'lime', 'lavender', 'maroon', 'navy', 'olive', 'teal', 'aqua', 'fuchsia', 'gray', 'silver']
    
    # Adiciona os gráficos à figura para cada coluna numérica
    for i, col in enumerate(num_cols, start=1):
        # Define a cor usando a lógica de ciclagem
        color = colors[i % len(colors)]
        
        # Adiciona o histograma
        fig.add_trace(go.Histogram(x=df[col], nbinsx=40, marker_color=color, 
                                    showlegend=False),
                      row=i, col=1)
        
        # Adiciona o boxplot
        fig.add_trace(go.Box(y=df[col], marker_color=color, showlegend=False),
                      row=i, col=2)
    
    # Configura o layout
    fig.update_layout(height=300*len(num_cols), title_text="Distribuição e Boxplot para cada variável numérica")
    fig.show()


# Importando o Dataset

In [3]:
df = pd.read_csv('MKT.csv')
df

youtube  facebook  newspaper  sales
0      84.72     19.20      48.96  12.60
1     351.48     33.96      51.84  25.68
2     135.48     20.88      46.32  14.28
3     116.64      1.80      36.00  11.52
4     318.72     24.00       0.36  20.88
..       ...       ...        ...    ...
166    45.84      4.44      16.56   9.12
167   113.04      5.88       9.72  11.64
168   212.40     11.16       7.68  15.36
169   340.32     50.40      79.44  30.60
170   278.52     10.32      10.44  16.08

[171 rows x 4 columns]

# Etapas de Desenvolvimento

Para te ajudar nesse processo, detalhar o processo nas etapas a seguir:

## Etapa 01) Análise Descritiva
Esta etapa consiste em explorar os dados do dataset para **compreender melhor as variáveis e identificar problemas**. Para isso, é recomendado utilizar a biblioteca **Pandas** para importar e manipular os dados e realizar cálculos estatísticos, além das bibliotecas de visualização.

É importante investigar o tipo de dado em cada variável, os valores e a distribuição dos
dados. Ao final, espera-se ter uma interpretação sólida dos dados para avançar para a
próxima etapa

### Análise Descritiva do Dataset

In [4]:
df_informations(df)

<H3 style='text-align:left;float:lfet;'>Informações sobre o Dataset

<H5> Dataset tem 171 linhas e 4 colunas. Não possui linhas duplicadas. Sobre o dataset, temos:

<H3 style='text-align:left;float:lfet;'>Sobre Dtypes, temos:

<H3 style='text-align:left;float:lfet;'>Estatística Descritiva

count   mean    std  min   1%   25%    50%    75%    99%    max
youtube   171.00 178.02 102.45 0.84 6.01 91.08 179.76 262.98 351.73 355.68
facebook  171.00  27.67  17.91 0.00 0.44 11.70  26.76  43.68  59.28  59.52
newspaper 171.00  35.24  24.90 0.36 1.75 13.74  31.08  50.88 103.42 121.08
sales     171.00  16.92   6.31 1.92 5.60 12.54  15.48  20.82  30.85  32.40

As informações **.describe** podem ser úteis para entender a dispersão e a tendência central dos dados.

- <font color='chartreuse'>Média</font>: Representa a tendência central dos dados;
- <font color='chartreuse'>Desvio Padrão</font>: Indica a dispersão, sugerindo uma variabilidade dos valores;
- <font color='chartreuse'>Quartis</font>: Fornecem informações sobre a distribuição dos dados ao longo de diferentes partes; e
- <font color='chartreuse'>Valor Mínimo e Máximo</font>: Indicam a faixa em que os dados estão concentrados.

Logo, podemos observar que:

1. **Variável "youtube":**
    * A média de investimento do YouTube é de aproximadamente 178.02, o que nos mostra que é o que mais recebe invetimento.
    * O desvio padrão é relativamente alto (102.45), o que sugere que os dados do YouTube apresentam uma dispersão considerável em relação à média.
    * Os quartis indicam que 25% dos dados estão abaixo de 91.08, 50% estão abaixo de 179.76 e 75% estão abaixo de 262.98.
    * O valor mínimo é de 0.84 e o valor máximo é de 355.68, mostrando que a faixa de visualizações varia de forma considerável.
<br></br>
1. **Variável "facebook":**
    * A média (27.67) de investimento do Facebook é aproximadamente 6.5x menor em relação ao Youtube.
    * O desvio padrão é de 17.91, indicando uma dispersão alta (64.72%) em relação à média.
    * Os quartis mostram que 25% dos dados estão abaixo de 11.70, 50% estão abaixo de 26.76 e 75% estão abaixo de 43.68.
    * O valor mínimo é 0.00 e o valor máximo é 59.52, mostrando a faixa de valores para a métrica do Facebook.
<br></br>
1. **Variável "newspaper":**
    * A média (35.24) do newspaper é a segunda maior ficando atrás apenas do Youtube.
    * O desvio padrão é de 24.90, indicando uma dispersão relativamente alta (70.65%) em relação à média.
    * Os quartis mostram que 25% dos dados estão abaixo de 13.74, 50% estão abaixo de 31.08 e 75% estão abaixo de 50.88.
    * O valor mínimo é 0.36 e o valor máximo é 121.08, indicando uma ampla faixa de valores.
<br></br>
1. **Variável "sales":**
    * A média para os dados de vendas é de aproximadamente 16.92.
    * O desvio padrão é de 6.31, sugerindo uma dispersão moderada (37.29%) em relação à média.
    * Os quartis mostram que 25% dos dados estão abaixo de 12.54, 50% estão abaixo de 15.48 e 75% estão abaixo de 20.82.
    * O valor mínimo é 1.92 e o valor máximo é 32.40, mostrando a faixa de valores para as vendas.

### Análise Gráfica

In [5]:
plot_distribution_and_boxplot(df)

## Etapa 02) Análise Exploratória
Neta etapa iremos explorar mais a fundo os dados, **identificando relações entre as variáveis e descobrindo padrões relevantes**. Para isso, utilize técnicas de visualização de dados e análises estatísticas, buscando possíveis correlações e identificando possíveis outliers ou desvios da normalidade.

### Correlação

#### Definição de Correlação

A correlação é uma medida estatística que *avalia a relação entre duas variáveis*. Ela indica a direção e a intensidade dessa relação, ou seja, se as variáveis se movem em conjunto (correlação positiva) ou de forma oposta (correlação negativa), e o quão forte essa relação é.

A força da correlação pode ser definida com base no valor do coeficiente de correlação, que varia de -1 a +1. Aqui estão as interpretações comuns para determinar a força da correlação:

1. <font color='yellow'>Correlação fraca:</font>
    * Quando o coeficiente de correlação está próximo de 0, a correlação é considerada fraca.
    * Seja ela positiva ou negativa, a relação entre as variáveis é considerada fraca se o coeficiente de correlação estiver próximo de zero (por exemplo, entre -0,3 e 0,3).
    * Nesse caso, as variáveis têm uma associação limitada e seus movimentos não são consistentes.
<br></br>
2. <font color='orange'>Correlação média:</font>
    * Uma correlação é considerada média quando o coeficiente de correlação está em torno de -0,5 a -0,3 ou de 0,3 a 0,5.
    * A relação entre as variáveis é moderada e mostra algum grau de consistência em seus movimentos.
    * Uma correlação média indica que as variáveis têm alguma influência mútua, mas não é uma relação forte.
<br></br>
3. <font color='chartreuse'>Correlação forte:</font>
    * A correlação é considerada forte quando o coeficiente de correlação está próximo de -1 ou 1.
    * Uma correlação positiva forte (próxima de +1) indica que as variáveis estão fortemente relacionadas e tendem a se mover na mesma direção.
    * Uma correlação negativa forte (próxima de -1) indica que as variáveis estão fortemente relacionadas, mas se movem em direções opostas.
    * Nesses casos, as variáveis têm uma associação consistente e os movimentos de uma variável estão altamente relacionados aos movimentos da outra.


É importante destacar que a força da correlação pode variar de acordo com o contexto e o domínio dos dados. Além disso, a correlação não implica causalidade direta, ou seja, apenas porque duas variáveis estão correlacionadas, não significa que uma causa a outra.

#### Matriz de Correlação

In [6]:
# Matriz de Correlação
correlation_matrix = df.corr()

# Cria o gráfico de heatmap
fig = go.Figure(data=go.Heatmap(
                   z=correlation_matrix,
                   x=correlation_matrix.columns,
                   y=correlation_matrix.columns,
                   colorscale='Viridis'))

# Adiciona anotações
annotations = []
for i, row in enumerate(correlation_matrix.values):
    for j, val in enumerate(row):
        annotations.append(go.layout.Annotation(text=str(round(val, 2)), x=correlation_matrix.columns[j], y=correlation_matrix.columns[i], 
                                                showarrow=False, font=dict(color='white')))

# Atualiza layout
fig.update_layout(title='Matriz de Correlação',
                  annotations=annotations,
                  height=500, width=600)

fig.show()


Explicado o que é correlação, podemos observar que:

1. A correlação entre "youtube" e "sales" é de 0.78, o que indica uma correlação positiva forte. Isso sugere que há uma forte relação entre o investimento no YouTube e as vendas. Aumento nas vendas tendem a estar associados ao aumento de investimento no YouTube.

2. A correlação entre "facebook" e "sales" é de 0.60, o que indica uma correlação positiva forte. Isso sugere que há uma relação forte entre o investimento no Facebook e as vendas. Aumentos nas vendas tendem a estar associados ao aumento de investimento no Facebook.

3. A correlação entre "newspaper" e "sales" é de 0.25, o que indica uma correlação positiva fraca. Isso sugere que há uma relação fraca entre o investimento em newspaper e as vendas. A influência dos newspaper nas vendas é limitada, pois a correlação é relativamente baixa.

#### Scatter plot: Sales vs Investimento

In [7]:
# Reestrutura os dados para um formato longo
df_melted = df.melt(id_vars='sales', var_name='platform', value_name='investment')

# Cria um dicionário com os símbolos para cada plataforma
symbols = {'youtube': 'circle', 'facebook': 'diamond', 'newspaper': 'square'}

# Cria o gráfico de dispersão
fig = px.scatter(df_melted, x='investment', y='sales', color='platform', 
                 symbol=df_melted['platform'].map(symbols), 
                 title='Scatter plot: Sales vs Investimento',
                 labels={'investment':'Investimento', 'sales':'Sales'}, 
                 hover_data=['platform', 'investment', 'sales'])

fig.show()


Analisando o Scatter plot é possível observar que é necessário desprender ou investir muito mais dinheiro no Youtube para ter o mesmo retorno que as demais plataformas. Em contra partida o Facebook e Newspaper possuem uma distribuição semelhante, onde a relação Sales vs Investimento é mais vantajosa, sendo necessário despender menos dinheiro para um retorno maior.

### Analisando a Eficiência da relação Sales vs Investimento

#### Cálculo do ROI individual

Uma forma de avaliar a eficiência dos investimentos em marketing é através do cálculo do Retorno sobre Investimento (ROI, em inglês Return on Investment). O ROI é uma métrica amplamente utilizada para avaliar a rentabilidade de um investimento. Ele é calculado da seguinte maneira:

**ROI = (Retorno do Investimento - Custo do Investimento) / Custo do Investimento**

Onde:

* Retorno do Investimento corresponde às Vendas (sales)
* Custo do Investimento é o valor investido nas plataformas (Youtube, Facebook e Newspaper)


Nossos dados indicam que as vendas (sales) estão correlacionadas com o valor investido em cada uma das plataformas. Ou seja, existe uma relação entre o valor investido e as vendas resultantes. Entretanto, é importante ressaltar que o valor total de vendas é influenciado pelo investimento em todas as plataformas, e não apenas em uma individualmente.

Embora calcular o ROI individual de cada plataforma possa não fornecer um quadro completamente preciso, pode ser uma maneira útil de começar a analisar a eficácia dos investimentos em marketing em cada plataforma.

Para simplificar, vamos fazer a suposição de que cada dólar investido em uma plataforma específica contribui igualmente para o total de vendas. Por exemplo, se investimos $84.72 no Youtube e $19.20 no Facebook, vamos considerar que cada um desses investimentos contribuiu igualmente para gerar as vendas de $12.60.

Com base nessas suposições, podemos prosseguir para analisar qual plataforma proporciona o melhor retorno sobre o investimento.

In [8]:
df['ROI_youtube'] = (df['sales'] - df['youtube']) / df['youtube']
df['ROI_facebook'] = (df['sales'] - df['facebook']) / df['facebook']
df['ROI_newspaper'] = (df['sales'] - df['newspaper']) / df['newspaper']
df

youtube  facebook  newspaper  sales  ROI_youtube  ROI_facebook   
0      84.72     19.20      48.96  12.60        -0.85         -0.34  \
1     351.48     33.96      51.84  25.68        -0.93         -0.24   
2     135.48     20.88      46.32  14.28        -0.89         -0.32   
3     116.64      1.80      36.00  11.52        -0.90          5.40   
4     318.72     24.00       0.36  20.88        -0.93         -0.13   
..       ...       ...        ...    ...          ...           ...   
166    45.84      4.44      16.56   9.12        -0.80          1.05   
167   113.04      5.88       9.72  11.64        -0.90          0.98   
168   212.40     11.16       7.68  15.36        -0.93          0.38   
169   340.32     50.40      79.44  30.60        -0.91         -0.39   
170   278.52     10.32      10.44  16.08        -0.94          0.56   

     ROI_newspaper  
0            -0.74  
1            -0.50  
2            -0.69  
3            -0.68  
4            57.00  
..             ...  
166          -0.45  
167           0.20  
168           1.00  
169          -0.61  
170           0.54  

[171 rows x 7 columns]

In [9]:
df.describe().T

count   mean    std   min   25%    50%    75%    max
youtube       171.00 178.02 102.45  0.84 91.08 179.76 262.98 355.68
facebook      171.00  27.67  17.91  0.00 11.70  26.76  43.68  59.52
newspaper     171.00  35.24  24.90  0.36 13.74  31.08  50.88 121.08
sales         171.00  16.92   6.31  1.92 12.54  15.48  20.82  32.40
ROI_youtube   171.00  -0.84   0.22 -0.96 -0.92  -0.90  -0.85   1.29
ROI_facebook  171.00    inf    NaN -0.96 -0.52  -0.32   0.32    inf
ROI_newspaper 171.00   0.41   4.70 -0.90 -0.68  -0.51   0.10  57.00

##### Corrigindo o inf do ROI_facebook

Após análise dos dados foi observado que na linha 98 do dataset a variável facebook é igual a 0, sendo assim quando calculamos o ROI, temos:

**ROI = (10.56-0)/0**

Matematicamente não é possível dividir um número por zero, então para corrigir esse problema vamos adotar uma premissa.

In [10]:
df.loc[98]

youtube         96.24
facebook         0.00
newspaper       11.04
sales           10.56
ROI_youtube     -0.89
ROI_facebook      inf
ROI_newspaper   -0.04
Name: 98, dtype: float64

Como premissa vamos adotar que o 0 será a média.

In [11]:
df['ROI_facebook'][98] = df['facebook'].mean()

In [12]:
df.describe().T

count   mean    std   min   25%    50%    75%    max
youtube       171.00 178.02 102.45  0.84 91.08 179.76 262.98 355.68
facebook      171.00  27.67  17.91  0.00 11.70  26.76  43.68  59.52
newspaper     171.00  35.24  24.90  0.36 13.74  31.08  50.88 121.08
sales         171.00  16.92   6.31  1.92 12.54  15.48  20.82  32.40
ROI_youtube   171.00  -0.84   0.22 -0.96 -0.92  -0.90  -0.85   1.29
ROI_facebook  171.00   0.62   3.48 -0.96 -0.52  -0.32   0.32  28.00
ROI_newspaper 171.00   0.41   4.70 -0.90 -0.68  -0.51   0.10  57.00

O ROI negativo, isso geralmente significa que o custo do investimento excede o retorno que você obteve. No contexto de uma campanha de marketing, isso poderia significar que o dinheiro que você gastou na campanha é maior do que a receita gerada por essa campanha.

Por exemplo, com um ROI médio de -0.90, se o valor investido foi de R$100, a perda média, de acordo com esse ROI, seria de R$90.

* **ROI_youtube:** ROI médio de **<font color='red'>-0.84</font>** sugere que, em média, para cada unidade de moeda investida, a perda foi de 84% do valor investido. Essa é uma situação financeira **<font color='red'>muito ruim</font>**, indicando que o investimento está perdendo dinheiro em vez de gerar lucro.
  
* **ROI_facebook:** ROI médio de **<font color='chartreuse'>0.62</font>** sugere que, em média, para cada unidade de moeda investida, o ganho foi de 62% do valor investido. Essa é uma situação financeira **<font color='chartreuse'>muito boa</font>**, indicando que o investimento está gerando lucro.
  
* **ROI_newspaper:** ROI médio de **<font color='chartreuse'>0.41</font>** sugere que, em média, para cada unidade de moeda investida, o ganho foi de 41% do valor investido. Essa é uma situação financeira **<font color='chartreuse'>muito boa</font>**, indicando que o investimento está gerando lucro.


#### Cálculo do ROI consolidado

Primeiramente vamos precisar somar o valor investido em todas as plataformas.

In [13]:
df['plataformas'] = df['youtube'] + df['facebook'] + df['newspaper']
df

youtube  facebook  newspaper  sales  ROI_youtube  ROI_facebook   
0      84.72     19.20      48.96  12.60        -0.85         -0.34  \
1     351.48     33.96      51.84  25.68        -0.93         -0.24   
2     135.48     20.88      46.32  14.28        -0.89         -0.32   
3     116.64      1.80      36.00  11.52        -0.90          5.40   
4     318.72     24.00       0.36  20.88        -0.93         -0.13   
..       ...       ...        ...    ...          ...           ...   
166    45.84      4.44      16.56   9.12        -0.80          1.05   
167   113.04      5.88       9.72  11.64        -0.90          0.98   
168   212.40     11.16       7.68  15.36        -0.93          0.38   
169   340.32     50.40      79.44  30.60        -0.91         -0.39   
170   278.52     10.32      10.44  16.08        -0.94          0.56   

     ROI_newspaper  plataformas  
0            -0.74       152.88  
1            -0.50       437.28  
2            -0.69       202.68  
3            -0.68       154.44  
4            57.00       343.08  
..             ...          ...  
166          -0.45        66.84  
167           0.20       128.64  
168           1.00       231.24  
169          -0.61       470.16  
170           0.54       299.28  

[171 rows x 8 columns]

Depois, calcular o ROI consolidado.

In [14]:
df['ROI'] = (df['sales'] - df['plataformas']) / df['plataformas']
df

youtube  facebook  newspaper  sales  ROI_youtube  ROI_facebook   
0      84.72     19.20      48.96  12.60        -0.85         -0.34  \
1     351.48     33.96      51.84  25.68        -0.93         -0.24   
2     135.48     20.88      46.32  14.28        -0.89         -0.32   
3     116.64      1.80      36.00  11.52        -0.90          5.40   
4     318.72     24.00       0.36  20.88        -0.93         -0.13   
..       ...       ...        ...    ...          ...           ...   
166    45.84      4.44      16.56   9.12        -0.80          1.05   
167   113.04      5.88       9.72  11.64        -0.90          0.98   
168   212.40     11.16       7.68  15.36        -0.93          0.38   
169   340.32     50.40      79.44  30.60        -0.91         -0.39   
170   278.52     10.32      10.44  16.08        -0.94          0.56   

     ROI_newspaper  plataformas   ROI  
0            -0.74       152.88 -0.92  
1            -0.50       437.28 -0.94  
2            -0.69       202.68 -0.93  
3            -0.68       154.44 -0.93  
4            57.00       343.08 -0.94  
..             ...          ...   ...  
166          -0.45        66.84 -0.86  
167           0.20       128.64 -0.91  
168           1.00       231.24 -0.93  
169          -0.61       470.16 -0.93  
170           0.54       299.28 -0.95  

[171 rows x 9 columns]

In [15]:
df.describe().T

count   mean    std   min    25%    50%    75%    max
youtube       171.00 178.02 102.45  0.84  91.08 179.76 262.98 355.68
facebook      171.00  27.67  17.91  0.00  11.70  26.76  43.68  59.52
newspaper     171.00  35.24  24.90  0.36  13.74  31.08  50.88 121.08
sales         171.00  16.92   6.31  1.92  12.54  15.48  20.82  32.40
ROI_youtube   171.00  -0.84   0.22 -0.96  -0.92  -0.90  -0.85   1.29
ROI_facebook  171.00   0.62   3.48 -0.96  -0.52  -0.32   0.32  28.00
ROI_newspaper 171.00   0.41   4.70 -0.90  -0.68  -0.51   0.10  57.00
plataformas   171.00 240.93 112.17 25.68 146.82 248.40 337.74 520.32
ROI           171.00  -0.92   0.02 -0.97  -0.94  -0.93  -0.91  -0.84

ROI médio Consolidado de **<font color='red'>-0.92</font>** sugere que, em média, para cada unidade de moeda investida, a perda foi de 92% do valor investido. Essa é uma situação financeira **<font color='red'>muito ruim</font>**, indicando que o investimento está perdendo dinheiro em vez de gerar lucro.

### Outliers

#### Definição de Outliers

Outliers são valores em um conjunto de dados que são significativamente diferentes dos outros. Eles são valores extremos que se desviam da média e das medidas padrão do conjunto de dados.

Para identificar outliers em um conjunto de dados, você pode usar vários métodos. Aqui estão algumas opções populares:

1. **Gráficos Boxplot**
   
2. **Z-Score**
   
3. **Desvio absoluto mediano (MAD)**
   
4. <font color='orange'>**Regra do IQR (Interquartile Range)**</font>

A análise de outliers é crucial para uma análise de dados precisa, pois outliers podem distorcer os resultados. Boxplots são uma ferramenta gráfica comum para identificar outliers, baseando-se no Intervalo Interquartil (IQR).

Apesar da facilidade de identificar outliers visualmente em boxplots, elaborei uma função para calcular numericamente o IQR e os limites de outliers. Essa função nos permite entender melhor o processo de detecção de outliers.

Apesar de útil, não é necessário calcular manualmente esses valores sempre. Ferramentas como boxplots fazem isso automaticamente, facilitando a identificação de outliers. Na sequência, utilizaremos boxplots para visualizar a presença de outliers em nossos dados.

O IQR é a diferença entre o terceiro quartil (Q3) e o primeiro quartil (Q1) dos dados. Qualquer ponto de dados que esteja abaixo de <font color='orange'>**Q1-1.5IQR**</font> ou acima de <font color='orange'>**Q3+1.5IQR**</font> pode ser considerado um outlier.

#### Cálculo dos Outliers

In [16]:
def calculate_iqr(df):
    # Define um DataFrame vazio para armazenar os resultados
    iqr_df = pd.DataFrame(columns=['Coluna', 'Q1', 'Q3', 'IQR', 'Lower Bound', 'Upper Bound', 'Min', 'Max'])
    
    for col in df.columns:
        if np.issubdtype(df[col].dtype, np.number):  # Se a coluna for numérica
            Q1 = df[col].quantile(0.25)
            Q3 = df[col].quantile(0.75)
            IQR = Q3 - Q1
            lower_bound = Q1 - 1.5 * IQR
            upper_bound = Q3 + 1.5 * IQR
            min_val = df[col].min() # valor mínimo
            max_val = df[col].max() # valor máximo
            # Cria um dicionário temporário com os dados
            temp_dict = {'Coluna': col, 'Q1': Q1, 'Q3': Q3, 'IQR': IQR, 
                         'Lower Bound': lower_bound, 'Upper Bound': upper_bound, 
                         'Min': min_val, 'Max': max_val}
            # Cria um DataFrame temporário a partir do dicionário
            temp_df = pd.DataFrame([temp_dict])
            # Concatena este DataFrame temporário com o DataFrame principal
            iqr_df = pd.concat([iqr_df, temp_df], ignore_index=True)
    return iqr_df

# Usando a função no DataFrame df
iqr_df = calculate_iqr(df[['youtube', 'facebook', 'newspaper', 'sales']])
iqr_df


Coluna    Q1     Q3    IQR  Lower Bound  Upper Bound  Min    Max
0    youtube 91.08 262.98 171.90      -166.77       520.83 0.84 355.68
1   facebook 11.70  43.68  31.98       -36.27        91.65 0.00  59.52
2  newspaper 13.74  50.88  37.14       -41.97       106.59 0.36 121.08
3      sales 12.54  20.82   8.28         0.12        33.24 1.92  32.40

Os outliers são valores que desviam notavelmente do restante dos dados. No contexto do IQR, outliers são aqueles valores que estão **abaixo** do "Lower Bound" ou **acima** do "Upper Bound". Assim, na tabela IQR que geramos, qualquer valor da coluna "Min" menor que o respectivo "Lower Bound" e qualquer valor da coluna "Max" maior que o respectivo "Upper Bound" são considerados outliers.

In [17]:
def plot_boxplots(df):
    # Calcula o número de linhas para os subplots
    num_rows = math.ceil(len(df.columns) / 2)

    # Cria a figura com subplots
    fig = make_subplots(rows=num_rows, cols=2)

    # Adiciona os boxplots à figura para cada coluna numérica
    for i, col in enumerate(df.columns, start=1):
        fig.add_trace(
            go.Box(
                y=df[col], 
                name=col, 
                boxmean=True, 
                showlegend=False
            ), 
            row=(i-1)//2 + 1, 
            col=((i-1)%2) + 1
        )

    # Atualiza o layout da figura
    fig.update_layout(height=400*num_rows, title_text="Boxplot de cada coluna numérica")

    # Mostra a figura
    fig.show()

# Usando a função com nosso DataFrame df
plot_boxplots(df[['youtube', 'facebook', 'newspaper', 'sales']])


Ao analisar os boxplots das nossas variáveis 'youtube', 'facebook', 'newspaper' e 'sales', notamos que apenas a variável 'newspaper' apresenta outliers. No entanto, essa quantidade de outliers parece ser pequena e, portanto, é improvável que cause um impacto significativo em nossa análise posterior ou distorça nossos resultados.

### Conclusão Etapa 2 - Análise Exploratória

1. **Correlação:** A análise de correlação indicou que o investimento em YouTube e Facebook tem uma forte relação positiva com as vendas, com o YouTube mostrando a relação mais forte. Isso sugere que aumentar o investimento nessas plataformas poderia resultar em um aumento nas vendas. Por outro lado, o investimento em Newspaper mostrou uma relação mais fraca com as vendas, indicando que essa plataforma pode não ser tão eficaz para impulsionar as vendas.

2. **Sales vc Investimento:** Ao analisar a relação de investimento vs vendas através dos gráficos de dispersão, notamos que o investimento no YouTube requer mais recursos para alcançar o mesmo nível de vendas comparado ao Facebook e Newspaper. Isso indica que o Facebook e Newspaper podem oferecer um retorno maior por unidade de investimento.

3. **Eficiência - ROI Individual:** O cálculo do ROI por plataforma revelou uma situação preocupante para o YouTube, com um ROI médio de -0.84. Isso sugere que o investimento no YouTube está atualmente gerando uma perda em vez de lucro. Em contrapartida, o Facebook e o Newspaper mostraram um ROI positivo, indicando que esses investimentos estão gerando lucro.

4. **Eficiência - ROI Consolidado:** Entretanto, ao considerar o ROI consolidado de todos os investimentos, vemos que a situação geral é de perda, com um ROI de -0.92. Isso indica que o investimento total em marketing está gerando uma perda, o que pode requerer uma reavaliação da estratégia de marketing.

5. **Outliers:** Por último, a análise de outliers revelou a existência de alguns outliers apenas na variável Newspaper. Embora a presença de outliers possa impactar algumas análises estatísticas, a quantidade observada é pequena e, portamente, é pouco provável que altere significativamente nossas conclusões.

## Etapa 03) Modelagem
Para esta etapa, deve-se **construir um modelo** simples de **regressão** que permita a previsão solicitada pela empresa, com base nos dados disponíveis. Para isto, importe as bibliotecas necessárias e carregue os conjuntos de dados para iniciar a sua construção!

### Separando em conjuntos de Treino e Teste

In [21]:
# Dividindo o conjunto de dados
X = df[['youtube', 'facebook', 'newspaper']]
y = df['sales']

# Dividindo os dados em conjuntos de treino e teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

# Verificando os tamanhos dos conjuntos de treino e teste
print("Treino:", X_train.shape, y_train.shape)
print("Teste:", X_test.shape, y_test.shape)

Treino: (136, 3) (136,)
Teste: (35, 3) (35,)


#### Treinando o modelo de Regressão Linear

In [22]:
# Criando e treinando o modelo de regressão linear
model = LinearRegression()
model.fit(X_train, y_train)

# Previsão do conjunto de teste
y_pred = model.predict(X_test)

#### Avaliação do Modelo

In [23]:
# Avaliação do modelo
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

print("RMSE: ", rmse)
print("R2: ", r2)

RMSE:  2.361522863329236
R2:  0.872863085701216


In [27]:
# Crie um novo dataframe df_full
df_full = df.copy()

# Antes da divisão, crie uma nova coluna chamada 'split'
df_full['split'] = 'train'

# Divida os dados
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Indique que os dados em X_test estão na partição de teste
df_full.loc[X_test.index, 'split'] = 'test'

# Treine o modelo
model = LinearRegression()
model.fit(X_train, y_train)

# Crie as previsões
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Adicione as previsões ao dataframe
df_full.loc[X_train.index, 'prediction'] = y_train_pred
df_full.loc[X_test.index, 'prediction'] = y_test_pred

# Agora crie o scatter plot
fig = px.scatter(
    df_full, x='sales', y='prediction',
    marginal_x='histogram', marginal_y='histogram',
    color='split', trendline='ols'
)
fig.update_traces(histnorm='probability', selector={'type':'histogram'})
fig.add_shape(
    type="line", line=dict(dash='dash'),
    x0=y.min(), y0=y.min(),
    x1=y.max(), y1=y.max()
)

fig.show()


## Etapa 04) Calculando Predição
Para concluirmos a demanda solicitada pela empresa, iremos **aplicar o modelo de regressão construído** nas etapas anteriores **para realizar as previsões** de retorno de vendas que pode ser gerado a partir de um determinado investimento em publicidade e assim, poderemos apresentá-lo a empresa.

Através dessas previsões, poderemos avaliar o impacto dos diferentes níveis de investimento em marketing nas vendas, auxiliando na tomada de decisões e na definição de estratégias de negócio.